R/loadFromMothur.R

```diff
@@ -0,0 +1,206 @@
+    if (!is.null(taxonomyTableFile)) {
+      taxa_tab <- .read_mothur_taxonomy(taxonomyTableFile, feature_tab, ...)
+    } else {
+      taxa_tab <- S4Vectors:::make_zero_col_DataFrame(nrow(feature_tab))
+    }
+    
+    # If colData information exists, gets that. Otherwise, sample_tab is just data frame without information
+    if (!is.null(sampleMetaFile)) {
+      sample_meta <- .read_mothur_sample_meta(sampleMetaFile, feature_tab)
+    } else {
+      sample_meta <- S4Vectors:::make_zero_col_DataFrame(ncol(feature_tab))
+    }
+    
+    if (!is.null(phyTreeFile)) {
+      tree <- .read_qza(phyTreeFile, ...)
+    } else {
+      tree <- NULL
+    }
+    
+    # if row.names(feature_tab) is a DNA sequence,  set it as refseq
+    if (!is.null(refSeqFile)){
+      refseq <- .read_qza(refSeqFile, ...)
+    } else if (featureNamesAsRefSeq) {
+      refseq <- .rownames_as_dna_seq(rownames(feature_tab))
+    } else {
+      refseq <- NULL
+    }
+    
+    return(TreeSummarizedExperiment(
+      assays = S4Vectors::SimpleList(counts = feature_tab),
+      rowData = taxa_tab,
+      colData = sample_meta,
+      rowTree = tree,
+      referenceSeq = refseq
+    ))
+    
+}
+
+.read_mothur_feature <- function(featureTableFile, ...){
+
+    # Reads the file
+    assay <- read.table(featureTableFile, check.names=FALSE, header=TRUE,
+                      sep="\t", stringsAsFactors=FALSE)
+  
+    # File contains additional columns "label", "numOtus", and "Group". They are removed
+    assay$label <- NULL
+    assay$numOtus <- NULL
+    # Information of "Group" column is saved, it includes taxa information
+    x <- assay$Group
+    assay$Group <- NULL
+    
+    # Initializes rownames
+    rownames(assay) <- NULL
+    # Saves taxa to rownames
+    rownames(assay) <- x
+    
+    # Transposes and converts assay from dataframe to matrix
+    assay <- t(as.matrix(assay))
+    
+    return(assay)
+    
+}
+
+.read_mothur_taxonomy <- function(taxonomyTableFile, feature_tab, ...){
+    
+    # Reads the file
+    rowData <- read.table(taxonomyTableFile, check.names=FALSE,
+                          header=TRUE, sep="\t", stringsAsFactors=FALSE)
+    # Deletes additional information between taxa levels.
+    rowData$Taxonomy <- gsub("[\"]", "", rowData$Taxonomy)
+    rowData$Taxonomy <- gsub("[(1-100)]", "", rowData$Taxonomy)
+    # Separate taxa to own columns
+    rowData <- tidyr::separate(rowData, "Taxonomy", into=c("Kingdom",
+                                                           "Phylum", "Order", "Class", "Family", "Genus"), sep=";",
+                               extra="merge")
+    # Deletes ";" from the end of Genus names
+    rowData$Genus <- gsub(";", "", rowData$Genus)
+    # Deletes addition "Size" column
+    rowData$Size <- NULL
+    
+    # Adds taxa to rownames
+    rownames(rowData) <- rowData$OTU
+    # Deletes addition "OTU" column
+    rowData$OTU <- NULL
```

Comment:
This section is a bit weird in the sense that the data stored as `rownames` here must match the `rownames` of the assay, if they are set. From the code this is not clearly managed which is the actual true feature name in the end returned by `rownames(tse)`.

Reply:
So, should the `rownames(tse) `be set somehow?  I assumed that it is checked that `rownames(assay) == rownames(rowData)`. Otherwise, I think it throws an error when TSE is created: `"rowdata and assay are not parallel" `or something

Reply:
Exactly. So it is critical, that you have a consistant strategy. If I remember correctly, assay names take precedence, when creating a `SE`.

Currently, you are doing the following steps:

- get assay data
- set the colnames of the `SE` of the assay as rownames, but then transpose the matrix
- get the rowData and set the rownames of the `SE` as rownames of the rowData
- get colData and just keep the columns which match the rownames, ... ah sorry the colnames, of the assay

So the strategy works for sure, but is not really consistent, since you are using three different types of using the dimnames

Good practice would be:

- [x] to extract the assay at a single point
- [x] to extract the colnames at a single point
- [ ] to extract the rowData at a single point
- [x] to extract the rownames at a single point
- [ ] to extract the colData at a single point
- [ ] to combine the data and set the dimnames in a single step

Especially the subsetting/ordering of the colData leaves me with a weird feeling, since in the worst case scenario, `NA` get produced without the user getting notified, that the data is not matching. The second thing is the assay and dimname setting. I would do something along the lines of:

```r
MOTHUR_NON_ASSAY_COLS <- c("label","numOtus","Group")
data <- read.table(featureTableFile, check.names=FALSE, header=TRUE,
                      sep="\t", stringsAsFactors=FALSE)
if(){
... # test for colnames not NULL
... # test for colnames present
}
assay <- as.matrix(data[,!(colnames() %in% MOTHUR_NON_ASSAY_COLS )])
assay <- t(assay)
if(){
... # test for rownames not NULL
}
colData <- DataFrame(data[,MOTHUR_NON_ASSAY_COLS])
return(list(assay = assay,
         colData = colData))
```

This way you don't touch any data, you don't have to and it is really easy to change, since you only have to change `MOTHUR_NON_ASSAY_COLS` to affect the outcome.

Reply:
For the rowData extraction, I would do something similar:

```r
 # Reads the file
MOTHUR_TAX_COL <- "Taxonomy"
data <- read.table(taxonomyTableFile, check.names=FALSE,
                      header=TRUE, sep="\t", stringsAsFactors=FALSE)
if(){
... # test for colnames present
... # check for MOTHUR_TAX_COL  present in colnames
}
# split taxa level into seperate columns
data [,MOTHUR_TAX_COL] <- gsub("[\"]", "", data [,MOTHUR_TAX_COL])
data [,MOTHUR_TAX_COL] <- gsub("[(1-100)]", "", data [,MOTHUR_TAX_COL])
tax <- tidyr::separate(data , MOTHUR_TAX_COL, into=c("Kingdom", "Phylum", "Order", "Class", "Family", "Genus"), sep=";", extra="merge")
tax$Genus <- gsub(";", "", tax$Genus)

# 
rowData <- cbind(DataFrame(tax),
                DataFrame(data[, !(colnames(data) %in% MOTHUR_TAX_COL)]))
if(!is.null(rowData$OTU)){
    if(){
    ... # check if the rownames are the same as in assay.
    }
    rownames(rowData) <- rowData$OTU
}
return(rowData)
```

Here again, you don't touch anything you don't need to.

The